### PR TITLE
Add isomorphic card abstraction to universal poker

### PR DESCRIPTION
**Pull request by aliyakamercan**
I added isomorphic card abstraction using https://github.com/kdub0/hand-isomorphism, currently it only supports 52 card deck (4 suits, 13 ranks).

You can use it by passing `cardAbstraction` GameParameter when creating the game:
```python
game = pyspiel.load_game("universal_poker",
                         {"gamedef": pyspiel.GameParameter(universal_poker_leduc_limit_2p),
                          "cardAbstraction": pyspiel.GameParameter("isomorphic")})
```

Default setting is no abstraction. 

`ObservationString`, `ObservationTensor`, `InformationStateString`, and `InformationStateTensor` are now using `AbstractedHoleAndBoardCards`. Didn't touch other places where `(Hole|Board)Cards` are used.

Future work:
- Modify original repo to support non-standard decks
- Support chaining card abstractions
- Would be nice to "register" abstractions w/o modifying the library code

### CODE REVIEW

File: open_spiel/games/universal_poker.cc [hunk 1]
```diff
@@ -709,19 +735,6 @@ int UniversalPokerGame::MaxGameLength() const {
  */
 std::string UniversalPokerGame::parseParameters(const GameParameters &map) {
   if (map.find("gamedef") != map.end()) {
-    // We check for sanity that all parameters are empty
-    if (map.size() != 1) {
```
Comment (aliyakamercan) [inline open_spiel/games/universal_poker.cc:713]: Had to remove this to allow passing `cardAbstraction` along with `gamedef`. 

The other option is to pass everything in `gamedef` as a single game parameter which is cumbersome. I think this is how `bettingAbstraction` is supposed to be used as well?

**Comment by lanctot** [comment]
Hi @aliyakamercan, thanks for this. Just a heads-up that we still plan to look into it.. just got busy with vacations and an influx of small PRs. Planning to get to this in the coming weeks. Apologies for the delay!

File: open_spiel/games/universal_poker.cc (hunk 1 above)
Comment (finbarrtimbers) [reply open_spiel/games/universal_poker.cc:713]: I think this is fine as-is.

Comment (finbarrtimbers) [reply open_spiel/games/universal_poker.cc:713]: i.e. removing it is fine.

**Comment by lanctot** [comment]
Tagging @jhtschultz only because he's brought this up before, may have though about it!

File: open_spiel/games/universal_poker.cc [hunk 2]
```diff
@@ -606,6 +611,27 @@ UniversalPokerGame::UniversalPokerGame(const GameParameters &params)
...
+  std::string card_abstraction =
+      ParameterValue<std::string>("cardAbstraction");
+  if (card_abstraction == "isomorphic") {
+    if (acpc_game_.NumSuitsDeck() != 4 || acpc_game_.NumRanksDeck() != 13) {
+        SpielFatalError(
+            "Isomorphic card abstraction is only supported standard 52 card deck.");
+    }
+    std::vector<int> cards_per_round;
+    int i;
+    for( i = 0; i < acpc_game_.NumRounds(); ++i ) {
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker.cc:624]: Please initialize in the for loop.

File: open_spiel/games/universal_poker.cc [hunk 3]
```diff
@@ -606,6 +611,27 @@ UniversalPokerGame::UniversalPokerGame(const GameParameters &params)
...
+        cards_per_round.push_back(acpc_game_.GetNbBoardCardsRequired(i));
+    }
+    cards_per_round[0] += cards_per_round[0] + acpc_game_.GetNbHoleCardsRequired();
+    card_abstraction_ = new card_abstraction::IsomorphicCardAbstraction(cards_per_round);
+  } else if (card_abstraction == "noop"){
+    card_abstraction_ = new card_abstraction::NoopCardAbstraction();
+  } else {
+    SpielFatalError(absl::StrFormat("cardAbstraction: %s not supported.",
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker.cc:632]: Nit: Can you make this `%s`? This is useful to detect erroneous whitespace.

File: open_spiel/games/universal_poker.cc (hunk 1 above)
Comment (finbarrtimbers) [reply open_spiel/games/universal_poker.cc:713]: Actually, no, can you modify to check that the only other field is `cardAbstraction`?

File: open_spiel/games/universal_poker.h [hunk 4]
```diff
@@ -160,6 +166,7 @@ class UniversalPokerState : public State {
   std::string actionSequence_;
 
   BettingAbstraction betting_abstraction_;
+  card_abstraction::CardAbstraction *card_abstraction_;
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker.h:169]: Can you use CardAbstraction* instead?

File: open_spiel/games/universal_poker.h [hunk 5]
```diff
@@ -189,6 +199,7 @@ class UniversalPokerGame : public Game {
   const acpc_cpp::ACPCGame acpc_game_;
   absl::optional<int> max_game_length_;
   BettingAbstraction betting_abstraction_ = BettingAbstraction::kFULLGAME;
+  card_abstraction::CardAbstraction *card_abstraction_;
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker.h:202]: Ditto.

File: open_spiel/games/universal_poker/card_abstraction/isomorphic.cc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+extern "C" {
+#include "open_spiel/games/universal_poker/hand-isomorphism/src/hand_index.h"
+}
+
+namespace open_spiel::universal_poker::card_abstraction {
+
+IsomorphicCardAbstraction::IsomorphicCardAbstraction(
+    std::vector<int> cards_per_round) {
+
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker/card_abstraction/isomorphic.cc:11]: Can you add a call to `reserve` here? We know the number of rounds.

File: open_spiel/games/universal_poker/card_abstraction/isomorphic.cc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+uint8_t from_iso_card(uint8_t iso_card) {
+    return deck_get_suit(iso_card) + 4 * deck_get_rank(iso_card);
+}
+
+std::pair<logic::CardSet, logic::CardSet>
+IsomorphicCardAbstraction::abstract(logic::CardSet hole_cards,
+                                    logic::CardSet board_cards) const {
+    auto h_arr = hole_cards.ToCardArray();
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker/card_abstraction/isomorphic.cc:34]: Please don't use auto; I don't know what the type is.

File: open_spiel/games/universal_poker/card_abstraction/isomorphic.cc [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    auto b_arr = board_cards.ToCardArray();
+    h_arr.insert(h_arr.end(), b_arr.begin(), b_arr.end());
+
+    int total_cards = h_arr.size();
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker/card_abstraction/isomorphic.cc:38]: const int

File: open_spiel/games/universal_poker/card_abstraction/isomorphic.cc [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    for(int i = 0; i < total_cards; i++) {
+        cards[i] = to_iso_card(h_arr[i]);
+    }
+
+    while (total_cards > 0) {
+        total_cards -= cards_per_round_[++round];
+    }
+
+    auto indexer = indexers_[round];
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker/card_abstraction/isomorphic.cc:50]: Please be explicit about the type.

File: open_spiel/games/universal_poker/card_abstraction/isomorphic.cc [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    auto index = hand_index_last(&indexer, cards);
+    hand_unindex(&indexer, round, index, cards);
+
+    logic::CardSet r_hole_cards;
```
Comment (finbarrtimbers) [inline open_spiel/games/universal_poker/card_abstraction/isomorphic.cc:54]: Please be explicit about the type.

File: open_spiel/scripts/install.sh [hunk 11]
```diff
@@ -105,9 +105,10 @@ fi
 # This Github repository contains the raw code from the ACPC server
 # http://www.computerpokercompetition.org/downloads/code/competition_server/project_acpc_server_v1.0.42.tar.bz2
 # with the code compiled as C++ within a namespace.
-DIR="open_spiel/games/universal_poker/acpc"
+DIR="open_spiel/games/universal_poker"
```
Comment (finbarrtimbers) [inline open_spiel/scripts/install.sh:108]: Why are you changing this? (I'm fine changing it, but what's the reason?)

File: open_spiel/scripts/install.sh [hunk 12]
```diff
@@ -105,9 +105,10 @@ fi
...
 if [[ ${BUILD_WITH_ACPC:-"ON"} == "ON" ]] && [[ ! -d ${DIR} ]]; then
-  git clone -b 'master' --single-branch --depth 1  https://github.com/jblespiau/project_acpc_server.git ${DIR}
+  git clone -b 'master' --single-branch --depth 1  https://github.com/jblespiau/project_acpc_server.git ${DIR}/acpc
+  git clone -b 'master' --single-branch --depth 1  https://github.com/aliyakamercan/hand-isomorphism.git ${DIR}/hand-isomorphism
```
Comment (finbarrtimbers) [inline open_spiel/scripts/install.sh:111]: Can you change the license for your project to Apache? I don't think we're allowed to import code with non-standard licenses. 

I can reach out to Kevin to change his license, as he works at DeepMind.

File: open_spiel/games/universal_poker.cc [hunk 13]
```diff
@@ -606,6 +611,27 @@ UniversalPokerGame::UniversalPokerGame(const GameParameters &params)
     SpielFatalError(absl::StrFormat("bettingAbstraction: %s not supported.",
                                     betting_abstraction));
   }
+
+  std::string card_abstraction =
+      ParameterValue<std::string>("cardAbstraction");
+  if (card_abstraction == "isomorphic") {
+    if (acpc_game_.NumSuitsDeck() != 4 || acpc_game_.NumRanksDeck() != 13) {
+        SpielFatalError(
+            "Isomorphic card abstraction is only supported standard 52 card deck.");
```
Comment (jhtschultz) [inline open_spiel/games/universal_poker.cc:620]: Small grammar fix, perhaps "is only supported" -> "only supports".

File: open_spiel/scripts/install.sh (hunk 11 above)
Comment (aliyakamercan) [reply open_spiel/scripts/install.sh:108]: to be able to reuse `DIR` in git clones

**Comment by aliyakamercan** [comment]
Addressed all comments, pls lmk if anything else stands out.

**Comment by xgshark** [comment]
Why was this not merged into the last release?

**Comment by finbarrtimbers** [comment]
Hey, I haven't had time to merge it. This is a non-trivial PR to merge as it requires that import kdub0/hand-isomorphism internally, which takes some work, and I have had other priorities, as OpenSpiel is not my primary work.

**Comment by lanctot** [comment]
I would still like to get this in and can try to get it merged in the next month or two for the next release.

@aliyakamercan can you resolve the conflicts with master? There have been a few changes in universal_poker, should be straight forward.

**Pull request closed without merging**